GameOfRunes/GameOfRunes/GameEngine/Components/HealthComponent.swift

```diff
@@ -10,10 +10,28 @@ import SpriteKit
 import GameplayKit
 
 class HealthComponent: GKComponent {
-    var healthPoints: Int
+    private weak var healthBarNode: HealthBarNode?
+    private var _healthPoints: Int
+    // if healthBarNode exists, use values from there, else use from _healthPoints
```

Comment:
Hmm, I did consider it. The potential issue is that player mana contains more information such as max amount of mana, max amount of health. This information is captured in the bar nodes, and thus I thought that the true state for player health/mana should be in the bar nodes while the true state for enemy health/mana is in the _health/mana.

Reply:
yeah so all these information should be contained in the ecs as well imo, since I'm thinking of moving all the nodes to sprite component instead and have the components perform calls to sprite system for updating instead

Reply:
Hmm but in enemy entity, the health component shouldnt have max health. Also in dropped mana entity, it shouldnt contain manaPerUnit and number of unit mana.

Reply:
The `HealthComponent` just needs to store something like initial health that acts as the max health and for `ManaComponent` it just needs to store max mana and current mana, then on update it sends information to the node to update it, that's all right? The rest of the information like `manaPerUnit` and number of unit mana should be information stored in the node because it's visual stuff then for powerups we just need to dictate how much mana is used and reflect it back on to the screen as bars -> 1 bar doesnt have to be 1 unit of mana in the underlying representation

Reply:
Hmm, it is possible. But we would still need to find some way to retrieve the `manaPerUnit` information either from meta-data or from the bar node as the power-up cost all calculated in terms of unit and not raw mana points.

Reply:
hey sorry i'm a bit confused by this convo. Isn't the HealthComponent part of ecs? So where exactly are u suggesting the state to be stored? Cos i agree that in master, different states are stored in different parts of the ecs, which is a little messy 

If i didn't interpret wrongly, you guys are suggesting that Component be the one that stores the state, and not the node right..?

Just want to clarify so that my future code keeps to this structure :)

Reply:
Agree with dom that it's probably a better to store this information in the ECS instead. If we go ahead with that, then maybe it's a better idea to separate player health and enemy health components? Seems like they have a few properties that differ (and would be nil if still using the same component)